File: java/p2/src/p2/types/operator/Assign.java
package p2.types.operator;

import java.util.Set;

import p2.lang.plan.Variable;
import p2.types.basic.Schema;
import p2.types.basic.Tuple;
import p2.types.basic.TupleSet;
import p2.types.exception.P2RuntimeException;
import p2.types.function.TupleFunction;
import p2.core.Runtime;

public class Assign extends Operator {
	
	private p2.lang.plan.Assignment assignment;
	
	private Schema schema;
	
	public Assign(Runtime context, p2.lang.plan.Assignment assignment, Schema input) {
		super(context, assignment.program(), assignment.rule());
		this.assignment = assignment;
		this.schema = input.clone();
		
		if (!this.schema.contains(this.assignment.variable())) {
			this.schema.append(this.assignment.variable());
		}
	}
	
	@Override
	public String toString() {
		return this.assignment.toString();
	}

	@Override
	public TupleSet evaluate(TupleSet tuples) throws P2RuntimeException {
		Variable variable = assignment.variable();
		TupleFunction<Comparable> function = assignment.value().function();
		TupleSet deltas = new TupleSet(tuples.name());
		for (Tuple tuple : tuples) {
			Tuple delta = tuple.clone();
			delta.value(variable, function.evaluate(delta));
			deltas.add(delta);
		}
		return deltas;
	}
	
	@Override
	public Schema schema() {
		return this.schema;
	}

	@Override
	public Set<Variable> requires() {
		return this.assignment.requires();
	}
	
}

File: lincoln/ruby/lib/types/operator/assign.rb
class Assign < Operator
  def initialize (assignment, input)
		super(assignment.program, assignment.rule)
		@assignment = assignment
		@schema = input.clone

		#@schema.append(@assignment.variable) unless @schema.includes? @assignment.variable 
		@schema << @assignment.variable unless @schema.contains @assignment.variable 
	end
	
	def to_s
		@assignment.to_s
	end

  def evaluate(tuples)
		variable = @assignment.variable
		function = @assignment.value.function
#		require 'ruby-debug'; debugger
		deltas = TupleSet.new(tuples.name)
		tuples.each do |tuple|
			delta = tuple.clone
			delta.set_value(variable, function.evaluate(delta))
			deltas << delta
		end
		return deltas
	end
	
	attr_reader :schema

	def requires
		@assignment.requires
	end
end
